### PR TITLE
Enable IDE0074: Use coalesce compound assignment

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

Use C# 8.0 [null-coalescing assignment](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/assignment-operator#null-coalescing-assignment).

## PR Context

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0054-ide0074

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/powershell/13396)

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/engine/interpreter/AddInstruction.cs [hunk 1]
```diff
@@ -130,14 +130,14 @@ public static Instruction Create(Type type)
...
-                case TypeCode.Single: return s_single ?? (s_single = new AddSingle());
-                case TypeCode.Double: return s_double ?? (s_double = new AddDouble());
+                case TypeCode.Int16: return s_int16 ??= new AddInt16();
+                case TypeCode.Int32: return s_int32 ??= new AddInt32();
+                case TypeCode.Int64: return s_int64 ??= new AddInt64();
+                case TypeCode.UInt16: return s_UInt16 ??= new AddUInt16();
+                case TypeCode.UInt32: return s_UInt32 ??= new AddUInt32();
+                case TypeCode.UInt64: return s_UInt64 ??= new AddUInt64();
+                case TypeCode.Single: return s_single ??= new AddSingle();
+                case TypeCode.Double: return s_double ??= new AddDouble();
```
Comment (vexx32) [inline src/System.Management.Automation/engine/interpreter/AddInstruction.cs:140]: At some point we should go through and make these switch expressions, there's a few of these sorts of things kicking around.

**Review by vexx32 — APPROVED** [review]

File: src/System.Management.Automation/engine/interpreter/AddInstruction.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/interpreter/AddInstruction.cs:140]: > make these switch expressions

In the code it seems it makes no sense.

Comment (vexx32) [reply src/System.Management.Automation/engine/interpreter/AddInstruction.cs:140]: There are many switches where every code path is a `return`; these could easily be
```csharp
type.GetTypeCode() switch
{
    TypeCode.Int16 => s_int16 ??= new AddInt16(),
    //...
}
```

Etc :)

Comment (xtqqczze) [reply src/System.Management.Automation/engine/interpreter/AddInstruction.cs:140]: @vexx32 Thanks, I was already considering the `IDE0066: Use switch expression` code fix for next PR.

**Comment by xtqqczze** [comment]
Codacy issue is not new.

**Comment by xtqqczze** [comment]
@TravisEz13 can you please review?

**Comment by daxian-dbw** [comment]
@PowerShell/powershell-maintainers reviewed this PR and agreed to accept it for 7.2-preview, namely delay merging this PR until after 7.1-RC1 release branch gets created. The decision is based on the following reason(s):
> The changes make sense, but they are actual code changes and the chance of regression is relatively high. So it requires line-by-line review to make sure changes are correct.

**Comment by iSazonov** [comment]
Set to Draft till 7.2 Preview1.

**Comment by iSazonov** [comment]
@xtqqczze Is it still Draft?

**Comment by xtqqczze** [comment]
rebased to restart tests

**Review by TravisEz13 — COMMENTED** [review]
Reviewed 11 out of 55 files.

**Comment by TravisEz13** [comment]
@PoshChan Please remind me in 7 days

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Comment by xtqqczze** [comment]
Resolved merge conflicts.

**Comment by iSazonov** [comment]
@TravisEz13 Do you want to continue the code review?

**Comment by xtqqczze** [comment]
@iSazonov Can we merge?

**Comment by xtqqczze** [comment]
@iSazonov CodeFactor is showing 3542 fixed issues for [this PR](https://www.codefactor.io/repository/github/powershell/powershell/commit-fixed/0d1ea73e599021c9ead87542646f7975bc04e5b9) taking us from grade F to grade D. Unfortunately the fixed issues list is not accurate :(

**Comment by iSazonov** [comment]
@xtqqczze Perhaps it is because the PR was old.

You could ask them https://www.codefactor.io/contact or twitter.

**Comment by xtqqczze** [comment]
> @xtqqczze Perhaps it is because the PR was old.
> 
> You could ask them https://www.codefactor.io/contact or twitter.
> 
> > **[Contact](https://www.codefactor.io/contact)**Quality software, faster.

Issue opened: codefactor-io/codefactor#27.

**Comment by xtqqczze** [comment]
> @xtqqczze we have prioritized this ticket and it is currently Work-In-Progress. Some updates have already been pushed to production so you should already see improved analysis results for C# 8.0 and 9.0. The full scope will be rolled out by the end of January 2021.

_Originally posted by @kaskas in https://github.com/codefactor-io/codefactor/issues/27#issuecomment-746392870_

**Pull request merged**